src/main/java/net/minecraftforge/fluids/ForgeFlowingFluid.java

```diff
@@ -0,0 +1,243 @@
+    private final Supplier<FlowingFluidBlock> block;
+    private final FluidAttributes.Builder builder;
+    private final boolean canMultiply;
+    private final int slopeFindDistance;
+    private final int levelDecreasePerBlock;
+    private final float explosionResistance;
+    private final BlockRenderLayer renderLayer;
+    private final int tickRate;
+
+    protected ForgeFlowingFluid(Properties properties)
+    {
+        this.flowing = properties.flowing;
+        this.still = properties.still;
+        this.builder = properties.attributes;
+        this.canMultiply = properties.canMultiply;
+        this.bucket = properties.bucket;
+        this.block = properties.block;
+        this.slopeFindDistance = properties.slopeFindDistance;
+        this.levelDecreasePerBlock = properties.levelDecreasePerBlock;
+        this.explosionResistance = properties.explosionResistance;
+        this.renderLayer = properties.renderLayer;
+        this.tickRate = properties.tickRate;
+    }
+
+    @Override
+    public Fluid getFlowingFluid()
+    {
+        return flowing.get();
+    }
+
+    @Override
+    public Fluid getStillFluid()
+    {
+        return still.get();
+    }
+
+    @Override
+    protected boolean canSourcesMultiply()
+    {
+        return canMultiply;
+    }
+
+    @Override
+    protected void beforeReplacingBlock(IWorld worldIn, BlockPos pos, BlockState state)
+    {
+        TileEntity tileentity = state.getBlock().hasTileEntity() ? worldIn.getTileEntity(pos) : null;
+        Block.spawnDrops(state, worldIn.getWorld(), pos, tileentity);
+    }
+
+    @Override
+    protected int getSlopeFindDistance(IWorldReader worldIn)
+    {
+        return slopeFindDistance;
+    }
+
+    @Override
+    protected int getLevelDecreasePerBlock(IWorldReader worldIn)
+    {
+        return levelDecreasePerBlock;
+    }
+
+    @Override
+    public BlockRenderLayer getRenderLayer()
+    {
+        return renderLayer;
+    }
+
+    @Override
+    public Item getFilledBucket()
+    {
+        return bucket != null ? bucket.get() : Items.AIR;
+    }
+
+    @Override
+    protected boolean func_215665_a(IFluidState p_215665_1_, IBlockReader p_215665_2_, BlockPos p_215665_3_, Fluid p_215665_4_, Direction p_215665_5_)
+    {
+        return p_215665_5_ == Direction.DOWN && !p_215665_4_.isIn(FluidTags.WATER);
+    }
+
+    @Override
+    public int getTickRate(IWorldReader p_205569_1_)
+    {
+        return tickRate;
+    }
+
+    @Override
+    protected float getExplosionResistance()
+    {
+        return explosionResistance;
+    }
+
+    @Override
+    protected BlockState getBlockState(IFluidState state)
+    {
+        if (block != null)
+            return block.get().getDefaultState().with(FlowingFluidBlock.LEVEL, getLevelFromState(state));
+        return Blocks.AIR.getDefaultState();
+    }
+
+    @Override
+    public boolean isEquivalentTo(Fluid fluidIn) {
+        return fluidIn == still.get() || fluidIn == flowing.get();
```

Comment:
Should this also compare any tags that this fluid is in and the given fluid is in? Or is that something that should be in an overwritten method.

Reply:
Although it may be interesting for a fluid to be able to say eg. that it's "just water" and it should be considered water by any neighbouring water blocks, such as for the level calculations, I don't think it would be correct to just compare tags, since there can be tags that mean something completely different, and we can't tell the difference in code.
I say if someone wants a different test than default, they can override isEquivalentTo.